src/polybench/POLYBENCH_ADI-Sycl.cpp

```diff
@@ -0,0 +1,169 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2017-24, Lawrence Livermore National Security, LLC
+// and RAJA Performance Suite project contributors.
+// See the RAJAPerf/LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include "POLYBENCH_ADI.hpp"
+
+#include "RAJA/RAJA.hpp"
+
+#if defined(RAJA_ENABLE_SYCL)
+
+#include "common/SyclDataUtils.hpp"
+
+#include <iostream>
+
+namespace rajaperf
+{
+namespace polybench
+{
+
+template < size_t work_group_size >
+void POLYBENCH_ADI::runSyclVariantImpl(VariantID vid)
+{
+  const Index_type run_reps = getRunReps();
+
+  auto res{getSyclResource()};
+  auto qu = res.get_queue();
+
+  POLYBENCH_ADI_DATA_SETUP;
+
+  if ( vid == Base_SYCL ) {
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      for (Index_type t = 1; t <= tsteps; ++t) {
+
+        const size_t global_size = work_group_size * RAJA_DIVIDE_CEILING_INT(n-2, work_group_size);
+
+        qu->submit([&] (sycl::handler& h) {
+          h.parallel_for(sycl::nd_range<1> (global_size, work_group_size),
+                         [=] (sycl::nd_item<1> item) {
+
+            Index_type i = item.get_global_id(0) + 1;
+
+            if (i < n-1) {
+              POLYBENCH_ADI_BODY2;
+              for (Index_type j = 1; j < n-1; ++j) {
+                 POLYBENCH_ADI_BODY3;
+              }
+              POLYBENCH_ADI_BODY4;
+              for (Index_type k = n-2; k >= 1; --k) {
+                 POLYBENCH_ADI_BODY5;
+              }
+            }
+
+          });
+        });
+
+        qu->submit([&] (sycl::handler& h) {
+          h.parallel_for(sycl::nd_range<1> (global_size, work_group_size),
+                         [=] (sycl::nd_item<1> item) {
+
+            Index_type i = item.get_global_id(0) + 1;
+
+            if (i < n-1) {
+              POLYBENCH_ADI_BODY6;
+              for (Index_type j = 1; j < n-1; ++j) {
+                 POLYBENCH_ADI_BODY7;
+              }
+              POLYBENCH_ADI_BODY8;
+              for (Index_type k = n-2; k >= 1; --k) {
+                 POLYBENCH_ADI_BODY9;
+              }
+            }
+
+          });
+        });
+
+      }  // tstep loop
+
+    }
+    stopTimer();
+
+  } else if (vid == RAJA_SYCL) {
+
+    POLYBENCH_ADI_VIEWS_RAJA;
+
+    using EXEC_POL =
+      RAJA::KernelPolicy<
+        RAJA::statement::SyclKernelAsync<
+          RAJA::statement::For<0, RAJA::sycl_global_0<work_group_size>,
```

Comment:
In kernel do we always use a 3d grid, if so should we use `sycl_global_2` here?

Reply:
Maybe we could set the sycl indexing dimension as an argument to SyclKernelAsync.

Reply:
I'm mainly concerned because in cuda and hip the z dimension doesn't always support the same range of values as x and y.

Reply:
As I understand it, a key difference between CUDA/HIP and SYCL is how thread indexing works.

Suppose we have a block or work-group with shape (nx, ny, nz) and an element with id (x, y, z). 
   -- In CUDA, the index is x + y * nx + z * nx * ny  ("column-major")
   -- In SYCL, the index is z + y * nz + x * nz * ny  ("row-mjor")

So CUDA/HIP threadIdx.x/y/z --> sycl::nd_item.get_local_id(2/1/0). The "2" index is the fastest for SYCL. Similarly:

gridDim.x/y/z  --> sycl::nd_item.get_group_range(2/1/0)
blockIdx.x/y/z --> sycl::nd_item.get_group(2/1/0)
blockDim.x/y/z --> sycl::nd_item.get_local_range().get(2/1/0)

Reply:
I'm trying to follow that pattern.

Reply:
ok, looks good now.